[Feature] Support data ingestion for range distribution table

srlch opened the pull request:
## What I'm doing:
Implement end-to-end range-based routing for range distribution tables on both FE and BE sides:

`FE`:
Add TabletRange and Tuple.toThrift / Variant.toThrift to serialize tablet ranges as TTabletRange/TVariant. OlapTableSink / FrontendServiceImpl now populate range_distributed_columns and per-tablet range in TOlapTablePartitionParam and TOlapTableIndexTablets, and temporarily disable colocate MV index for RANGE distribution tables.

`BE`:
Introduce RangeRouter to route rows to tablets based on TTabletRange, supporting multi-column ranges and various open/closed/±inf intervals; Extend TabletSinkSender with RangeTabletSinkSender.

`Tests`:
Add/extend FE tests (VariantTest, TabletRangeTest) to cover numeric, string, boolean and date/datetime serialization to Thrift. Add BE tests (RangeRouterTest, TabletSinkSenderRangeTest) covering typical routing paths, boundary behavior, sparse selections and error cases.

Fixes #64986

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Implements end-to-end range-distribution data routing, adding FE serialization and BE routing/components with protocol updates and extensive tests.
> 
> - **Protocol/IDL**:
>   - Introduce `TOlapTableDistributionType` and per-tablet `range` in `TOlapTableTablet`; switch `TVariant` to a single `value` field; add `TTabletRange`, `TTuple` helpers.
> - **Frontend (FE)**:
>   - Serialize `TabletRange`, `Tuple`, `Variant` to Thrift; include `distribution_type`, range distribution columns, and per-tablet `range` in `TOlapTablePartitionParam`/`TOlapTableIndexTablets`.
>   - `OlapTableSink`: populate new fields, disable colocate MV index for `RANGE` distribution.
>   - `SplitTabletJob`: set tablet `range` using `TabletRange`; fix base-index selection via `getBaseIndexMetaId`.
>   - `MetaUtils`: helpers to get range distribution columns/ids.
> - **Backend (BE)**:
>   - Add `RangeRouter` (parses/validates `TTabletRange`, routes via upper boundaries) and `RangeTabletSinkSender`; integrate in `OlapTableSink` when `is_range_distribution()`.
>   - Extend `OlapTablePartitionParam` to parse `distribution_type`; support hash/random; expose distribution slots; minor comparator overload.
>   - Make `_send_chunk_by_node` virtual.
> - **Tests/Build**:
>   - Add BE tests (`range_router_test`, `tablet_sink_sender_range_test`) and FE tests (`VariantTest`, `TabletRangeTest`, sink tests); update CMake lists.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f5c7e9315ec15dfdebd585bda4b4c9be34c5267c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

xiangguangyxg commented on fe/fe-core/src/main/java/com/starrocks/catalog/Variant.java:
```diff
@@ -104,4 +106,23 @@ public static int compatibleCompare(List<Variant> key1, List<Variant> key2) {
+            case INT:
+            case BIGINT:
+            case LARGEINT:
+                variant.setLong_value(getLongValue());
```
large int is int128, may exceed long range

srlch replied:
Done

xiangguangyxg commented on gensrc/thrift/Descriptors.thrift:
```diff
@@ -297,6 +297,8 @@ struct TOlapTablePartitionParam {
 
     9: optional bool enable_automatic_partition
+    // used for range distribution table to split data into multiple ranges
+    10: optional list<string> range_distributed_columns
```
```suggestion
    10: optional list<string> range_distribution_columns
```

xiangguangyxg commented on be/src/exec/tablet_sink_sender.cpp:
```diff
@@ -40,12 +40,83 @@ TabletSinkSender::TabletSinkSender(PUniqueId load_id, int64_t txn_id, IndexIdToT
+            const auto& row_indices = partition_row_indices_map[part_id];
+            // Get or create row-based range checker for this partition and index.
+            if (_range_checkers[index_schema->index_id].find(part_id) ==
+                _range_checkers[index_schema->index_id].end()) {
```
Better to not repeatedly retrieve an element from a map using [].

srlch replied:
Fixed

alvin-celerdata commented:
@cursor review

wyb commented on be/src/exec/range_router.h:
```diff
@@ -0,0 +1,60 @@
+class SlotDescriptor;
+
+// RangeRouter is responsible for routing rows will ranges which represent the entire (-inf, +inf)
+class RangeRouter {
```
can we reuse the partition range code?

alvin-celerdata commented:
@cursor review

xiangguangyxg commented on be/src/exec/range_router.h:
```diff
@@ -0,0 +1,60 @@
+    // the _boundaries will be: [100, 200]
+    // the _lower_bound_inclusive will be: [false, true, true]
+    std::vector<ColumnPtr> _boundaries;
+    std::vector<uint8_t> _lower_bound_inclusive;
```
[-inf, 100) this range is invalid, do not write ranges like this.

srlch replied:
Fix the incorrect comment

xiangguangyxg commented on be/src/exec/range_router.h:
```diff
@@ -0,0 +1,60 @@
+    // the _boundaries will be: [100, 200]
+    // the _lower_bound_inclusive will be: [false, true, true]
+    std::vector<ColumnPtr> _boundaries;
+    std::vector<uint8_t> _lower_bound_inclusive;
```
The first element of _lower_bound_inclusive is always false, so it do not need to save?

wyb commented on be/src/exec/tablet_info.cpp:
```
if (is_hash_distribution()) {
    // compute hashes
} else if (is_random_distribution()) {
    // rand Next
}
```

wyb replied:
`hashes` is not used in range distribution.

wyb commented on be/src/exec/range_router.cpp:
```diff
@@ -0,0 +1,311 @@
+            if (value.__isset.string_value) {
+                value_str = value.string_value;
+            } else if (value.__isset.long_value) {
+                value_str = std::to_string(value.long_value);
```
so only need string_value in proto?

srlch replied:
Yes, i have refactor the code and remove the long value part

xiangguangyxg commented on gensrc/proto/types.proto:
```diff
@@ -234,8 +234,7 @@ message TransactionStatePB {
     optional PTypeDesc type = 1;
-    optional int64 int_value = 2;
-    optional string string_value = 3;
+    optional string string_value = 2;
```
```suggestion
    optional string value = 2;
```

srlch replied:
Fixed

srlch commented on be/src/exec/range_router.cpp:
```diff
@@ -0,0 +1,311 @@
+        // However, we need to support configurations where there are multiple
+        // identical point ranges in the middle, e.g.:
+        //   (-inf, 100), [100, 100], [100, 100], (100, 200), [200, +inf)
+        // In that case, adjacent [100, 100] ranges are allowed.
```
We have to handle [100, 100] for duplicate table, because all the range keys for duplicate table are the same.

srlch commented on gensrc/thrift/Types.thrift:
```diff
@@ -632,8 +631,7 @@ struct TParquetOptions {
     1: optional TTypeDesc type
-    2: optional i64 int_value
-    3: optional string string_value
+    2: optional string string_value
```
Fixed

alvin-celerdata commented:
@cursor review

alvin-celerdata commented:
@cursor review

wyb commented on be/src/exec/range_router.cpp:
```diff
@@ -0,0 +1,239 @@
+
+    _upper_boundaries.resize(num_columns);
+    for (size_t i = 0; i < num_ranges - 1; ++i) {
+        _upper_boundaries_slice.emplace_back(&_upper_boundaries, i);
```
should check the ranges is ordered if use a vector

srlch replied:
`_validate_range` can ensure that

srlch committed:
[Feature] Support data ingestion for range distribution table

Implement end-to-end range-based routing for range distribution tables on both FE and BE sides:

FE:
Add TabletRange and Tuple.toThrift / Variant.toThrift to serialize tablet ranges as TTabletRange/TVariant.
OlapTableSink / FrontendServiceImpl now populate range_distributed_columns and per-tablet range in TOlapTablePartitionParam and TOlapTableIndexTablets, and temporarily disable colocate MV index for RANGE distribution tables.

BE:
Introduce RangeRouter to route rows to tablets based on TTabletRange, supporting multi-column ranges and various open/closed/±inf intervals;
Extend TabletSinkSender with RangeTabletSinkSender.

Tests:
Add/extend FE tests (VariantTest, TabletRangeTest) to cover numeric, string, boolean and date/datetime serialization to Thrift.
Add BE tests (RangeRouterTest, TabletSinkSenderRangeTest) covering typical routing paths, boundary behavior, sparse selections and error cases.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
fix

Signed-off-by: srlch <linzichao@starrocks.com>

alvin-celerdata commented:
@cursor review
